### PR TITLE
ARP for peer book entries

### PR DESCRIPTION
**Pull request by Vagabond**

### CODE REVIEW

File: src/group/libp2p_group_worker.erl [hunk 1]
```diff
@@ -509,7 +509,7 @@ kill_pid(undefined) ->
 kill_pid(Pid) ->
     erlang:unlink(Pid),
     erlang:exit(Pid, kill),
-    Pid.
+    undefined.
```
Comment (madninja) [inline src/group/libp2p_group_worker.erl]: This is an unrelated change. When I tried this in https://github.com/helium/erlang-libp2p/pull/238 it failed CT on my local machine, oddly for secure streams, where it previously passed

File: src/group/libp2p_group_gossip_server.erl [hunk 2]
```diff
@@ -99,14 +99,27 @@ handle_call(Msg, _From, State) ->
...
+handle_cast({handle_data, Key, Msg, StreamPid}, State=#state{}) ->
     %% Incoming message from a gossip stream for a given key
     case maps:find(Key, State#state.handlers) of
         error -> {noreply, State};
         {ok, {M, S}} ->
             %% Catch the callback response. This avoids a crash in the
             %% handler taking down the gossip_server itself.
-            catch M:handle_gossip_data(Msg, S),
+            try M:handle_gossip_data(Msg, S) of
+                {reply, Reply} ->
```
Comment (madninja) [inline src/group/libp2p_group_gossip_server.erl]: Why not allow the Key for the reply to be set too?

File: src/peerbook/libp2p_peer_resolution.erl [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+-export([resolve/3, install_handler/2]).
+
+%% Gossip group key to register and transmit with
+-define(GOSSIP_GROUP_KEY, "peer_resolution").
+
+-spec resolve(pid(), libp2p_crypto:pubkey_bin(), non_neg_integer()) -> ok.
+resolve(GossipGroup, PK, Ts) ->
+    libp2p_group_gossip:send(GossipGroup, ?GOSSIP_GROUP_KEY, 
+                             libp2p_peer_resolution_pb:encode_msg(#libp2p_peer_resolution_msg_pb{msg = {request, #libp2p_peer_request_pb{pubkey=PK, timestamp=Ts}}})),
```
Comment (madninja) [inline src/peerbook/libp2p_peer_resolution.erl]: indent long, long line

File: src/libp2p_transport_p2p.erl [hunk 4]
```diff
@@ -63,11 +64,17 @@ connect_to(MAddr, UserOptions, Timeout, TID) ->
...
                             end;
-                        {error, Error} -> {error, Error}
+                        {error, Error} ->
+                            libp2p_peerbook:refresh(Peerbook, PeerInfo),
+                            {error, Error}
                     end;
-                {error, Reason} -> {error, Reason}
+                {error, Reason} ->
+                    libp2p_peerbook:refresh(Peerbook, Addr),
+                    {error, Reason}
```
Comment (mrallen1) [inline src/libp2p_transport_p2p.erl]: Instead of repeating the refresh many times, maybe it would be better to handle it like this:

```erlang
  Response = case libp2p_peerbook:get(Peerbook, Addr) of
    %% ...
  end,
  case Response of
    {error, _Error} ->
      libp2p_peerbook:refresh(Peerbook, PeerInfo);
    _OK -> ok
  end,
  Response;
```

File: src/peerbook/libp2p_peer_resolution.erl (hunk 3 above)
Comment (mrallen1) [inline src/peerbook/libp2p_peer_resolution.erl]: Line wrap here please?

File: src/peerbook/libp2p_peer_resolution.erl [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+handle_gossip_data(Data, Handle) ->
+    case libp2p_peer_resolution_pb:decode_msg(Data, libp2p_peer_resolution_msg_pb) of
+        #libp2p_peer_resolution_msg_pb{msg = {request, #libp2p_peer_request_pb{pubkey=PK, timestamp=Ts}}} ->
+            %% look up our peerbook for a newer record for this peer
+            case libp2p_peerbook:get(Handle, PK) of
+                {ok, Peer} ->
+                    case libp2p_peer:timestamp(Peer) > Ts of
+                        true ->
+                            lager:notice("ARP response for ~p Success", [PK]),
+                            {reply, libp2p_peer_resolution_pb:encode_msg(#libp2p_peer_resolution_msg_pb{msg = {response, Peer}})};
```
Comment (mrallen1) [inline src/peerbook/libp2p_peer_resolution.erl]: Line wrap here too, please

File: src/group/libp2p_group_gossip_server.erl (hunk 2 above)
Comment (Vagabond) [reply src/group/libp2p_group_gossip_server.erl]: "don't cross the streams"
